presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.type.ArrayType;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.TimeZoneKey.getTimeZoneKey;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static java.util.Arrays.asList;
+
+public class TestArrayReduceFunction
+        extends AbstractTestFunctions
+{
+    public TestArrayReduceFunction()
+    {
+        super(testSessionBuilder().setTimeZoneKey(getTimeZoneKey("Pacific/Kiritimati")).build());
+    }
+
+    @Test
+    public void testNullOnly()
+            throws Exception
+    {
+        assertFunction("reduce(ARRAY [], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, null);
+    }
+
+    @Test
+    public void testBasic()
+            throws Exception
+    {
+        assertFunction("reduce(ARRAY [5, 20, 50], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, 75L);
+        assertFunction("reduce(ARRAY [5 + RANDOM(1), 20, 50], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, 75L);
+        assertFunction("reduce(ARRAY [5, 6, 10, 20], 0.0, (x, s) -> x + s, s -> s)", DOUBLE, 41.0);
+    }
+
+    @Test
+    public void testNulls()
+            throws Exception
+    {
+        assertFunction("reduce(ARRAY [NULL], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, null);
+        assertFunction("reduce(ARRAY [NULL, NULL, NULL], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, null);
+        assertFunction("reduce(ARRAY [5, NULL, 50], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, 55L);
```

Comment:
The semantics of reduce seem weird and inconsistent. In particular, I find it odd that:

`reduce(array[null], 1, (x, s) -> x + s, s -> s)` => `null`

but 

`reduce(array[null, 0], 1, (x, s) -> x + s, s -> s)` => `1`

In the first case, it's understandable why it would produce null, since it's equivalent to `1 + null`.

In the second case, it's should produce the same as `1 + null + 0`, which means it should also produce null if we're to be consistent.